README.md
[![Stories in Ready](https://badge.waffle.io/universityofderby/chef-weblogic.png?label=ready&title=Ready)](https://waffle.io/universityofderby/chef-weblogic)
# Weblogic Cookbook

## Scope
This cookbook installs Oracle Weblogic Server.
Oracle Java is also installed as a required dependency.
It does not install or manage any services.

## Requirements
* Chef 12 or higher
* Network accessible artifact repository

## Platform
* Centos 6
* Red Hat Enterprise 6

## Usage

### Resources
```
weblogic '10.3.6' do
  ownername 'weblogic'
  groupname 'weblogic_admin' 
end
```
### Attributes
```
node['common_artifact_repo'] = 'https://artifact.server/software'
```

### Common Artifact Repository
This cookbook uses a common artifact repository pattern. 
Artifacts urls will be generated in the following manner:

node['common_artifact_repo'] + vendor + product + version + file_as_named_by_vendor

eg. the above resource & attribute will attempt to retrieve artifacts from the following
locations:
```
https://artifact.server/software/oracle/weblogic/10.3.6/wls1036_generic.jar
```

Artifact urls can be overridden by specifying installer_url on the
resource.

## License and Author

* Author: Dan Webb (<d.webb@derby.ac.uk>)
* Author: Luke Bradbury (<luke.bradbury@derby.ac.uk>)

Copyright: 2015, University of Derby

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
weblogic Cookbook CHANGELOG
========================
v2.1.5
Fix group membership #19

v2.1.4
Fix silent file owner #17

v2.1.3
Add weblogic 10.3.5
Changes teste CentOS version to 6.7

v2.1.2 (2015-12-15)
Add weblogic 12.2.1

v2.1.0 (2015-11-20)
Deprecate v2.1.1 as per issue #10

v2.1.0 (2015-11-20)
Move to custom resources

v2.0.2 (2015-07-16)
Fix Berkshelf to point to the new supermarket universe endpoint (@smford22)

v2.0.1 (2015-05-27)
Fixed issue #3: removed reference to old cookbook name.  Weblogic 12.1.1/12.1.2 tests working.

v2.0.0 (2015-05-21)
Bumped version to work around conflict.

v1.0.0 (2015-05-14)
Added waffle.io badge.

v0.1.1 (2015-05-10)
Initial release.

This now uses the resource cookbook which is currently under active development.
https://github.com/chef-cookbooks/resource


test/shared/spec_helper.rb
require 'serverspec'
# Required by serverspec
set :backend, :exec

RSpec.configure do |c|
  c.before :all do
    c.path = '/sbin:/usr/sbin:/bin'
  end
end


spec/unit/recipes/default_spec.rb
require 'chefspec'
require 'chefspec/berkshelf'

describe 'weblogic::default' do
  before(:each) do
    stub_command('sudo -V').and_return('blarg')
  end

  let :chef_run do
    ChefSpec::Runner.new do |node|
      node.set['weblogic']['installer']['url'] = 'http://example.com/weblogic.jar'
      node.set['oracle']['weblogic']['path'] = '/sw/app/oracle/weblogic'
    end.converge(described_recipe)
  end

  it "should create the temporary directory if it isn't there" do
    expect(chef_run).to create_directory('/sw/app/temp')
  end

  it 'should create the oracle weblogic directory' do
    expect(chef_run).to create_directory('/sw/app/oracle/weblogic')
  end

  it 'should include java recipe' do
    expect(chef_run).to include_recipe('java::default')
  end

  it 'should install weblogic' do
    expect(chef_run).to run_execute('install')
  end
end


test/fixtures/cookbooks/weblogic-server/usr/metadata.rb
name 'weblogic-server-usr'
version '0.0.1'

depends 'weblogic'


resources/weblogic_server.rb
#
# Cookbook Name:: weblogic
# Resource:: weblogic-server
#
# Copyright (C) 2015 University of Derby
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

property :version, String, name_property: true
property :cache_path, String, default: ::File.join(Chef::Config[:file_cache_path], "weblogic-#{@version}")
property :silent_file, String, default:
  if Gem::Version.new(@version) < Gem::Version.new('12.1.2.0.0')
    'silent.xml'
  else
    'silent.rsp'
  end
property :silent_path, String, default: ::File.join(@cache_path, @silent_file)

property :silent_cmd, String, default{
  if Gem::Version.new(version) < Gem::Version.new('12.1.2.0.0')
    '-mode=silent -silent_xml='
  else
    '-silent -responseFile '
  end
}

property :component_paths, String, default{
  v = Gem::Version.new(version)
  v1211 = Gem::Version.new('12.1.1')
  v1212 = Gem::Version.new('12.1.2')
  case
  when v >= v1211 && v < v1212
    'WebLogic Server/Core Application Server|WebLogic Server/Administration Console|WebLogic Server/Configuration Wizard and Upgrade Framework|WebLogic Server/Web 2.0 HTTP Pub-Sub Server|WebLogic Server/WebLogic JDBC Drivers|WebLogic Server/Third Party JDBC Drivers|WebLogic Server/WebLogic Server Clients|WebLogic Server/Xquery Support|WebLogic Server/Server Examples'
  else
    'WebLogic Server/Core Application Server|WebLogic Server/Administration Console|WebLogic Server/Configuration Wizard and Upgrade Framework|WebLogic Server/Web 2.0 HTTP Pub-Sub Server|WebLogic Server/WebLogic JDBC Drivers|WebLogic Server/Third Party JDBC Drivers|WebLogic Server/WebLogic Server Clients|WebLogic Server/WebLogic Web Server Plugins|WebLogic Server/UDDI and Xquery Support|WebLogic Server/Server Examples'
  end
}
property :installer_url, String, default {
  node['common_artifact_repo'] +
    "/oracle/weblogic-server/#{version}/#{installer_file}"
}
property :installer_path, String, default: ::File.join(cache_path, installer_file)
property :installer_file, String, default {
  v = Gem::Version.new(version)
  v1036 = Gem::Version.new('10.3.6')
  v1211 = Gem::Version.new('12.1.1')
  v1212 = Gem::Version.new('12.1.2')
  v1213 = Gem::Version.new('12.1.3')
  case
  when v >= v1036 && v < v1211
    'wls1036_generic.jar'
  when v >= v1211 && v < v1212
    'wls1211_generic.jar'
  when v >= v1212 && v < v1213
    'wls_121200.jar'
  else
    'fmw_12.1.3.0.0_wls.jar'
  end
}

property :ownername, String, default: 'oracle'
property :groupname, String, default: 'dba'
property :home, String, default: "/opt/oracle/Middleware/weblogic-#{version}"
property :oracle_jdk, String, default: '7'
property :inventory_path, String, default: ::File.join(home, 'oraInventory')

action :create do
  node.default['java']['jdk_version'] = oracle_jdk
  node.default['java']['install_flavor'] = 'oracle'
  node.default['java']['oracle']['accept_oracle_download_terms'] = true
  include_recipe 'java'

  group groupname

  user ownername do
    gid groupname
  end

  node.default['oracle']['inventory']['group'] = groupname
  node.default['oracle']['inventory']['user'] = ownername

  if Gem::Version.new(version) >= Gem::Version.new('12.1.2.0.0')
    include_recipe 'oracle-inventory'
    group node['oracle']['inventory']['group'] do
      append true
      members ownername
    end
  end

  directory cache_path do
    mode 00775
    owner ownername
    group groupname
    recursive true
  end

  directory home do
    mode 00775
    owner ownername
    group groupname
    recursive true
  end

  remote_file installer_path do
    mode 00644
    owner ownername
    group groupname
    source installer_url
  end

  template silent_path do
    mode 00644
    owner owner
    group groupname
    source silent_file + '.erb'
    cookbook 'weblogic'
    variables(
      home: home,
      component_paths: component_paths
    )
  end

  execute "install weblogic server #{version}" do
    user ownername
    group groupname
    command "java -jar #{installer_path}  #{silent_cmd}#{silent_path}"
    only_if { Dir["#{home}/*"].empty? }
  end
end


resources/weblogic.rb
#
# Cookbook Name:: weblogic
# Resource:: weblogic-server
#
# Copyright (C) 2015 University of Derby
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
resource_name :weblogic
default_action :create
property :version, String, name_property: true
property :cache_path, String, default: lazy { ::File.join(Chef::Config[:file_cache_path], "weblogic-#{version}") }
property :silent_file, String, default: lazy {
  if Gem::Version.new(@version) < Gem::Version.new('12.1.2.0.0')
    'silent.xml'
  else
    'silent.rsp'
  end
}
property :silent_path, String, default: lazy { ::File.join(cache_path, silent_file) }

property :silent_cmd, String, default: lazy {
  if Gem::Version.new(version) < Gem::Version.new('12.1.2.0.0')
    '-mode=silent -silent_xml='
  else
    '-silent -responseFile '
  end
}

property :component_paths, String, default: lazy {
  v = Gem::Version.new(version)
  v1211 = Gem::Version.new('12.1.1')
  v1212 = Gem::Version.new('12.1.2')
  case
  when v >= v1211 && v < v1212
    'WebLogic Server/Core Application Server|WebLogic Server/Administration Console|WebLogic Server/Configuration Wizard and Upgrade Framework|WebLogic Server/Web 2.0 HTTP Pub-Sub Server|WebLogic Server/WebLogic JDBC Drivers|WebLogic Server/Third Party JDBC Drivers|WebLogic Server/WebLogic Server Clients|WebLogic Server/Xquery Support|WebLogic Server/Server Examples'
  else
    'WebLogic Server/Core Application Server|WebLogic Server/Administration Console|WebLogic Server/Configuration Wizard and Upgrade Framework|WebLogic Server/Web 2.0 HTTP Pub-Sub Server|WebLogic Server/WebLogic JDBC Drivers|WebLogic Server/Third Party JDBC Drivers|WebLogic Server/WebLogic Server Clients|WebLogic Server/WebLogic Web Server Plugins|WebLogic Server/UDDI and Xquery Support|WebLogic Server/Server Examples'
  end
}
property :installer_url, String, default: lazy {
  node['common_artifact_repo'] + "/oracle/weblogic-server/#{version}/#{installer_file}"
}
property :installer_path, String, default: lazy { ::File.join(cache_path, installer_file) }
property :installer_file, String, default: lazy {
  v = Gem::Version.new(version)
  v1035 = Gem::Version.new('10.3.5')
  v1036 = Gem::Version.new('10.3.6')
  v1211 = Gem::Version.new('12.1.1')
  v1212 = Gem::Version.new('12.1.2')
  v1213 = Gem::Version.new('12.1.3')
  v1221 = Gem::Version.new('12.2.1')
  v1222 = Gem::Version.new('12.2.2')

  case
  when v >= v1035 && v < v1036
    'wls1035_generic.jar'
  when v >= v1036 && v < v1211
    'wls1036_generic.jar'
  when v >= v1211 && v < v1212
    'wls1211_generic.jar'
  when v >= v1212 && v < v1213
    'wls_121200.jar'
  when v >= v1213 && v < v1221
    'fmw_12.1.3.0.0_wls.jar'
  when v >= v1221 && v < v1222
    'fmw_12.2.1.0.0_wls.jar'
  else
    ''
  end
}

property :ownername, String, default: 'oracle'
property :groupname, String, default: 'dba'
property :home, String, default: lazy { "/opt/oracle/Middleware/weblogic-#{version}" }
property :oracle_jdk, String, default: '7'
property :inventory_path, String, default: lazy { ::File.join(home, 'oraInventory') }

action :create do
  node.default['java']['jdk_version'] = oracle_jdk
  node.default['java']['install_flavor'] = 'oracle'
  node.default['java']['oracle']['accept_oracle_download_terms'] = true
  include_recipe 'java'

  group groupname do
    append true
  end

  user ownername do
    gid groupname
  end

  node.default['oracle']['inventory']['group'] = groupname
  node.default['oracle']['inventory']['user'] = ownername

  if Gem::Version.new(version) >= Gem::Version.new('12.1.2.0.0')
    include_recipe 'oracle-inventory'
    group node['oracle']['inventory']['group'] do
      append true
      members ownername
    end
  end

  directory cache_path do
    mode 00775
    owner ownername
    group groupname
    recursive true
  end

  directory home do
    mode 00775
    owner ownername
    group groupname
    recursive true
  end

  remote_file installer_path do
    mode 00644
    owner ownername
    group groupname
    source installer_url
  end

  template silent_path do
    mode 00644
    owner ownername
    group groupname
    source silent_file + '.erb'
    cookbook 'weblogic'
    variables(
      home: home,
      component_paths: component_paths
    )
  end

  execute "install weblogic server #{version}" do
    user ownername
    group groupname
    command "java -jar #{installer_path}  #{silent_cmd}#{silent_path}"
    only_if { Dir["#{home}/*"].empty? }
  end
end
